src/services/completions.ts

```diff
@@ -1667,134 +1690,75 @@ namespace ts.Completions {
-                if (resolvedModuleSymbol !== moduleSymbol && every(resolvedModuleSymbol.declarations, isNonGlobalDeclaration)) {
-                    pushSymbol(resolvedModuleSymbol, InternalSymbolName.ExportEquals, moduleSymbol, file, isFromPackageJson);
-                }
-
-                for (const symbol of typeChecker.getExportsAndPropertiesOfModule(moduleSymbol)) {
-                    // If this is `export { _break as break };` (a keyword) -- skip this and prefer the keyword completion.
-                    if (some(symbol.declarations, d => isExportSpecifier(d) && !!d.propertyName && isIdentifierANonContextualKeyword(d.name))) {
-                        continue;
-                    }
-
-                    pushSymbol(symbol, symbol.name, moduleSymbol, file, isFromPackageJson);
+        /** Mutates `symbols`, `symbolToOriginInfoMap`, and `symbolToSortTextMap` */
+        function collectAutoImports(resolveModuleSpecifiers: boolean) {
+            if (!shouldOfferImportCompletions()) return;
+            Debug.assert(!detailsEntryId?.data);
+            const start = timestamp();
+            const moduleSpecifierCache = host.getModuleSpecifierCache?.();
+            host.log?.(`collectAutoImports: starting, ${resolveModuleSpecifiers ? "" : "not "}resolving module specifiers`);
+            if (moduleSpecifierCache) {
+                host.log?.(`collectAutoImports: module specifier cache size: ${moduleSpecifierCache.count()}`);
+            }
+            const lowerCaseTokenText = previousToken && isIdentifier(previousToken) ? previousToken.text.toLowerCase() : "";
+            const exportInfo = codefix.getSymbolToExportInfoMap(sourceFile, host, program);
+            const packageJsonAutoImportProvider = host.getPackageJsonAutoImportProvider?.();
+            const packageJsonFilter = detailsEntryId ? undefined : createPackageJsonImportFilter(sourceFile, host);
+            exportInfo.forEach((info, key) => {
+                const symbolName = key.substring(0, key.indexOf("|"));
+                if (!detailsEntryId && isStringANonContextualKeyword(symbolName)) return;
+                const isCompletionDetailsMatch = detailsEntryId && some(info, i => detailsEntryId.source === stripQuotes(i.moduleSymbol.name));
+                if (isCompletionDetailsMatch || stringContainsCharactersInOrder(symbolName.toLowerCase(), lowerCaseTokenText)) {
+                    // If we don't need to resolve module specifiers, we can use any re-export that is importable at all
+                    // (We need to ensure that at least one is importable to show a completion.)
+                    const { moduleSpecifier, exportInfo } = resolveModuleSpecifiers
+                        ? codefix.getModuleSpecifierForBestExportInfo(info, sourceFile, program, host, preferences)
+                        : { moduleSpecifier: undefined, exportInfo: find(info, isImportableExportInfo) };

```

Comment:
This is the real crux of the PR. In `getSymbolsFromOtherSourceFileExports`, we deduplicated re-exports on the fly, since they don’t matter until we need module specifiers. But for import statement completions, we need to resolve module specifiers up front (for reasons discussed beginning around https://github.com/microsoft/TypeScript/issues/31658#issuecomment-738455502), so `getSymbolsFromOtherSourceFileExports` was utterly unusable.

Now, instead of tossing out re-exports, we let `codefix.getSymbolToExportInfoMap` group exports (the grouping strategy is discussed in an earlier review comment). Then, if we are being asked to eagerly resolve module specifiers, we can send all the re-exports in the group to `codefix.getModuleSpecifierForBestExportInfo`, and if we are just being asked to show an auto-import completion as we’ve always done, we simply pick the first usable export in the group, since they’re interchangeable until we need module specifiers. This lets us use 95% of the same code between import statement completions (the new feature in this PR) and regular auto imports.